### PR TITLE
copy skel

### PR DESCRIPTION
**Pull request by TeeSrisantitham**

### CODE REVIEW

File: skel/cmd/sonic/app.go [hunk 1]
```diff
@@ -10,6 +11,7 @@ import (
 	"github.com/openware/sonic/skel/config"
 	"github.com/openware/sonic/skel/handlers"
 	"github.com/openware/sonic/skel/models"
+	"github.com/otiai10/copy"
```
Comment (mod) [inline skel/cmd/sonic/app.go:14]: in this framework I'm against including too many dependencies

while copying we also need to replace the package name
from "github.com/openware/sonic/skel/models" to "NAME/models"

`import (
    "io/ioutil"
    "path/filepath"
    "os"
    "strings"
)

func copy(source, destination string) error {
    var err error = filepath.Walk(source, func(path string, info os.FileInfo, err error) error {
        var relPath string = strings.Replace(path, source, "", 1)
        if relPath == "" {
            return nil
        }
        if info.IsDir() {
            return os.Mkdir(filepath.Join(destination, relPath), 0755)
        } else {
            var data, err1 = ioutil.ReadFile(filepath.Join(source, relPath))
            if err1 != nil {
                return err1
            }
            return ioutil.WriteFile(filepath.Join(destination, relPath), data, 0777)
        }
    })
    return err
}`

**Review by mod — CHANGES_REQUESTED** [review]
Make sure to code copy in pkg utils first with test if possible

File: skel/cmd/sonic/app.go (hunk 1 above)
Comment (TeeSrisantitham) [reply skel/cmd/sonic/app.go]: I have a question.

Will we copy all files and directories from 'skel' to the destination with a new NAME or just models in 'skel' ? 
I think we shouldn't copy all files. We should make some template directory for it.

**Comment by TeeSrisantitham** [comment]
> Make sure to code copy in pkg utils first with test if possible

Ok I'm working on it.

File: skel/cmd/sonic/app.go (hunk 1 above)
Comment (dpatsora) [reply skel/cmd/sonic/app.go]: We have to move all the files inside `skel` folder and replace every import definition `openware/skel` with `NAME/skel`

**Comment by TeeSrisantitham** [comment]
> > Make sure to code copy in pkg utils first with test if possible
> 
> Ok I'm working on it.

https://github.com/openware/pkg/pull/8

File: scripts/generate.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
```
Comment (mod) [inline scripts/generate.sh:2]: let's inspire from rvm, nvm

imagine we call it svm # sonic version manager

`curl sonic.io/svm | bash`

clone repo + install bash function or bash script "svm"

then user can type in bash 

```
svm github.com/openware/appsonic

cp -r $HOME/.svm/skel appsonic
sed -i "s#github.com/openware/sonic/skel#$APP#g" appsonic/**/*.go
```

**Comment by TeeSrisantitham** [comment]
You can test by

```curl -ssL https://raw.githubusercontent.com/TeeSrisantitham/sonic/feature/copy-skel/scripts/install.sh | bash```

```svm install appsonic ```

File: scripts/install.sh [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  local INSTALL_DIR
+  INSTALL_DIR="$HOME/.svm"
+
+  if [ -f "$INSTALL_DIR/sonic.go" ]; then
+    echo "=> svm is already installed in $INSTALL_DIR, trying to update the script"
+    cd ~/.svm
+    git pull origin master
+  else
+    echo "=> Cloning svm as script to '$INSTALL_DIR'"
+    git clone --quiet --branch feature/copy-skel https://github.com/TeeSrisantitham/sonic.git  ~/.svm
```
Comment (TeeSrisantitham) [inline scripts/install.sh:64]: Have to change to master later

File: scripts/svm.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  APP="${1}"
+
+  echo "=> Installing ${APP} to ${NAME}"
+
+  case $APP in 
+    'github.com/openware/appsonic' | 'appsonic')
+      local NAME
+      NAME="${2:-appsonic}"
+      cp -r $HOME/.svm/skel ${NAME}
+      sed -i '' "s/github.com\/openware\/sonic\/skel/${NAME}/g" ${NAME}/**/*.go ${NAME}/go.mod
```
Comment (TeeSrisantitham) [inline scripts/svm.sh:68]: This line still not work. I will continue to find a solution.

@mod @dpatsora Do you have any suggestions?

Comment (TeeSrisantitham) [reply scripts/svm.sh]: I fixed

File: scripts/install.sh [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  local DETECTED_PROFILE
+  DETECTED_PROFILE=''
+
+  if [ -n "${BASH_VERSION-}" ]; then
+    if [ -f "$HOME/.bashrc" ]; then
+      DETECTED_PROFILE="$HOME/.bashrc"
+    elif [ -f "$HOME/.bash_profile" ]; then
+      DETECTED_PROFILE="$HOME/.bash_profile"
+    fi
+  elif [ -n "${ZSH_VERSION-}" ]; then
```
Comment (dpatsora) [inline scripts/install.sh:36]: let's do it as separate if

File: scripts/svm.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  IFS='/'
+  read -A ADDR <<< "$DESTINATION"
+  DIR=${ADDR[${#ADDR[@]}]}
+
+  echo "=> Creating ${DIR}"
+
+  cp -r $HOME/.svm/skel ${DIR}
+  sed -i -e "s|github.com/openware/sonic/skel|${DESTINATION}|g" ${DIR}/**/*.go ${DIR}/go.mod
+  git init -q ${DIR}
```
Comment (dpatsora) [inline scripts/svm.sh:60]: git commit -m "Initiali commit"
Maybe, you can add `git remote add ` ( check it in [gin-generator](https://github.com/openware/generator-gin/blob/master/generators/app/index.js))

File: scripts/svm.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
```
Comment (dpatsora) [inline scripts/svm.sh:1]: Remove .go-e files

File: scripts/install.sh (hunk 5 above)
Comment (TeeSrisantitham) [reply scripts/install.sh:36]: If you use ZSH
Run by this command
```curl -ssL https://raw.githubusercontent.com/TeeSrisantitham/sonic/feature/copy-skel/scripts/install.sh | zsh```

**Review by mod — APPROVED** [review]

**Pull request merged**